Allow full profile updates through the PATCH handler

jhrozek opened the pull request:
# Summary

We used to use a bespoke PATCH handler to update profiles but the PATCH
method only allowed changing alert and remediate. We can make the code
simpler and allow more updates by modifying the existing profile using
the UpdateMask and then just relaying the update to the existing
UpdateProfile method of the Profile service.

Fixes: #2971

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [x] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

There are unit tests but you can also test with curl:

Adding a rule to a profile:
```
curl -XPATCH -H "Authorization: Bearer $MINDER_BEARER_TOKEN"
'http://localhost:8080/api/v1/profile/ce746798-c30c-46f3-8a12-0654747b49ec?context.project=1900c9f6-ef0f-43e4-8af7-108beb72de0e&context.provider=github'
-d '{"name":"dependabot-enabled-profile", "labels":[], "buildEnvironment":[], "repository":[{"type":"dependabot_configured","params":null,"def":{"apply_if_file":"requirements.txt","package_ecosystem":"pypi","schedule_interval":"weekly"},"name":"dependabot_configured"},{"type":"dependabot_configured","params":null,"def":{"apply_if_file":"go.mod","package_ecosystem":"gomod","schedule_interval":"weekly"},"name":"dependabot_configured_go"}], "pullRequest":[], "remediate ":"on", "alert":"off", "type":"", "version":"", "displayName":"gah" }
```

Remove the rule back:
```
curl -XPATCH -H "Authorization: Bearer $MINDER_BEARER_TOKEN"
'http://localhost:8080/api/v1/profile/ce746798-c30c-46f3-8a12-0654747b49ec?context.project=1900c9f6-ef0f-43e4-8af7-108beb72de0e&context.provider=github'
-d '{"name":"dependabot-enabled-profile", "labels":[], "buildEnvironment":[], "repository":[{"type":"dependabot_configured","params":null,"def":{"apply_if_file":"requirements.txt","package_ecosystem":"pypi","schedule_interval":"weekly"},"name":"dependabot_configured"}], "pullRequest":[], "remediate":"on", "alert":"off", "type":"", "version":"", "displayName":"gah"}}'
```


# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

dmjb commented on internal/controlplane/handlers_profile.go:
```diff
@@ -547,71 +549,63 @@ func (s *Server) PatchProfile(ctx context.Context, ppr *minderv1.PatchProfileReq
 		return nil, status.Errorf(codes.Internal, "failed to get profile: %s", err)
 	}
 
-	// TEMPORARY HACK: Since we do not need to support this patch logic in the
```
We still need this check.

jhrozek replied:
I can add a test but wouldn't the update code just error out later?

dmjb commented on internal/controlplane/handlers_profile.go:
```diff
@@ -547,71 +549,63 @@ func (s *Server) PatchProfile(ctx context.Context, ppr *minderv1.PatchProfileReq
-	if err != nil {
-		return nil, util.UserVisibleError(codes.Internal, "error updating profile: %v", err)
-	}
+	for _, attr := range updateMask.Paths {
```
Subscription ID is one which we do not want to allow to be changed.

dmjb reviewed (changes requested):
Leaving this as "Request changes" until the question around editing bundle profiles is settled.

jhrozek commented:
> After talking to @jhrozek offline, I understand why that check can be deleted. I would like to see a test added for the bundled subscription ID case though.

I added a test and moved a bunch of the patch handler to profile service.

btw this PR is NOT important to merge at this moment

evankanderson commented on internal/profiles/service.go:
```diff
@@ -276,6 +291,63 @@ func (p *profileService) UpdateProfile(
+	ctx context.Context,
+	projectID uuid.UUID,
+	profileID uuid.UUID,
+	subscriptionID uuid.UUID,
```
Do we want to allow `PATCH` with subscriptions?  If not, could we drop this argument?

evankanderson replied:
(It looks like we pass `uuid.Nil` as the argument to `UpdateProfile` from the other gRPC endpoint, so it looks like we the PATCH-on-Subscription codepath is effectively dead...)

jhrozek replied:
probably not, you're right. We'll likely to only update subscription profiles through a replace, not patch.

evankanderson commented on internal/profiles/service.go:
```diff
@@ -276,6 +291,63 @@ func (p *profileService) UpdateProfile(
+	}
+
+	for _, attr := range updateMask.Paths {
+		oldReflect := oldProfilePb.ProtoReflect()
+		patchReflect := patchPb.ProtoReflect()
```
Do you want to do these inside the `for` loop?

jhrozek replied:
thanks, fixed

evankanderson commented on internal/profiles/service.go:
```diff
@@ -276,6 +291,63 @@ func (p *profileService) UpdateProfile(
+	updateMask *fieldmaskpb.FieldMask,
+	qtx db.Querier,
+) (*minderv1.Profile, error) {
+	oldProfilePb, err := getProfilePBFromDB(ctx, profileID, projectID, qtx)
```
Do you want to call `old` "base" here?  It's slightly confusing, because we go through the steps and then pass `old` to the Update method.

jhrozek replied:
thanks, renamed

evankanderson commented on internal/profiles/service.go:
```diff
@@ -276,6 +291,63 @@ func (p *profileService) UpdateProfile(
+}
+
+func patchProfilePb(oldProfilePb, patchPb *minderv1.Profile, updateMask *fieldmaskpb.FieldMask) {
+	// if there is no update mask, return the old profile. The grpc-rest gateway always sets the update mask
```
This is confusing, as it says "return", but there's no return value.  Maybe you mean "an empty field mask means apply no changes"?

evankanderson reviewed (approved):
Only nits; I _like_ the new code much better!  Thanks!

jhrozek commented:
@evankanderson thanks for the review, I fixed your comments in separate patches for easier review.
